DirectProgramming/DPC++FPGA/Tutorials/Features/ac_int/src/ac_int.cpp

```diff
@@ -0,0 +1,237 @@
+  buffer<MyInt28, 1> d_buf(&d, 1);
+  buffer<MyInt15, 1> e_buf(&e, 1);
+
+  q.submit([&](handler &h) {
+    accessor a_acc(a_buf, h, read_only);
+    accessor b_acc(b_buf, h, read_only);
+    accessor c_acc(c_buf, h, write_only, no_init);
+    accessor d_acc(d_buf, h, write_only, no_init);
+    accessor e_acc(e_buf, h, write_only, no_init);
+    h.single_task<BasicOpsAcInt>([=]() [[intel::kernel_args_restrict]] {
+      c_acc[0] = a_acc[0] + b_acc[0];
+      d_acc[0] = a_acc[0] * b_acc[0];
+      e_acc[0] = a_acc[0] / b_acc[0];
+    });
+  });
+}
+
+void TestShiftOps(queue &q, const MyInt14 &a, const MyInt14 &b, MyInt14 &c) {
+  buffer<MyInt14, 1> a_buf(&a, 1);
+  buffer<MyInt14, 1> b_buf(&b, 1);
+  buffer<MyInt14, 1> c_buf(&c, 1);
+
+  q.submit([&](handler &h) {
+    accessor a_acc(a_buf, h, read_only);
+    accessor b_acc(b_buf, h, read_only);
+    accessor c_acc(c_buf, h, write_only, no_init);
+    h.single_task<ShiftOps>([=]() [[intel::kernel_args_restrict]] {
+      MyInt14 temp = a_acc[0] << b_acc[0];
+      c_acc[0] = temp >> b_acc[0];
+    });
+  });
+}
+
+void TestEfficientShiftOps(queue &q, const MyInt14 &a, const MyUInt2 &b,
+                           MyInt14 &c) {
+  buffer<MyInt14, 1> a_buf(&a, 1);
+  buffer<MyUInt2, 1> b_buf(&b, 1);
+  buffer<MyInt14, 1> c_buf(&c, 1);
+
+  q.submit([&](handler &h) {
+    accessor a_acc(a_buf, h, read_only);
+    accessor b_acc(b_buf, h, read_only);
+    accessor c_acc(c_buf, h, write_only, no_init);
+    h.single_task<EfficientShiftOps>([=]() [[intel::kernel_args_restrict]] {
+      MyInt14 temp = a_acc[0] << b_acc[0];
+      c_acc[0] = temp >> b_acc[0];
+    });
+  });
+}
+
+MyInt14 TestBitAccess(queue &q, const MyInt14 &a) {
+  MyInt14 res;
+  buffer<MyInt14, 1> a_buf(&a, 1);
+  buffer<MyInt14, 1> res_buf(&res, 1);
+
+  q.submit([&](handler &h) {
+    accessor a_acc(a_buf, h, read_only);
+    accessor res_acc(res_buf, h, write_only, no_init);
+    h.single_task<BitAccess>([=]() [[intel::kernel_args_restrict]] {
+      // 0b1111101
+      MyInt7 temp = a_acc[0].slc<7>(3);
+
+      res_acc[0] = 0; // Must be initialized before being accessed by the bit
+                      // select operator `[]`. Using the `[]` operator on an
+                      // uninitialized `ac_int` variable is undefined behavior
+                      // and can give you unexpected results.
+
+      // 0 -> 0b1111101000
+      res_acc[0].set_slc(3, temp);
+
+      // 0b1111101000 -> 0b1111101111
+      res_acc[0][2] = 1;
+      res_acc[0][1] = 1;
+      res_acc[0][0] = 1;
+    });
+  });
+  return res;
+}
+
+int main() {
+#if defined(FPGA_EMULATOR)
+  ext::intel::fpga_emulator_selector device_selector;
+#else
+  ext::intel::fpga_selector device_selector;
+#endif
+
+  bool passed = true;
+
+  try {
+    queue q(device_selector, dpc_common::exception_handler);
+
+    constexpr int kVal1 = 1000, kVal2 = 2;
+
+    // Kernel `BasicOpsInt` contains native `int` type addition, multiplication,
```

Comment:
Should this kind of comment be repeated or emphasized in the declaration of these functions?   In general the ratio of comments to code in the example is very low.    You don't want comments that repeat what the code says, but you do want comments that explain the purpose of the code that is close by when you read the code.

Reply:
I feel I can remove these chunks of comments in `main()`, and add some simplified comments with `Test*` function above.